### PR TITLE
ACL Cache v2 for EMQ X R30

### PR DESCRIPTION
**Pull request by terry-xiaoyu**
### cache using process process dictionary
An improvement solution of https://github.com/emqtt/emqttd/pull/1732.

Remove the `ordered circular buffer`, using an FIFO `keys-queue` instead. 

The cache is still implemented using process dictionary, and each cache entry in the cache is of following format:
```
|-- CacheK --|-- CacheV --|
```
where:
- CacheK: `{Access :: publish | subscribe, Topic :: topic()}`
- CacheV: `{AclResult :: allow | deny, TimeNow :: timestamp()}`

The `CacheK` are also stored in the `keys-queue`, in the order of their insertion.

### cache replacement and eviction 
- If the max-size reached, the oldest cache entry will be replaced first, and then the 2nd oldest one. This is called `replacement` or `eviction`.
- It would try to remove all the expired cache entries if the result returned by `get_acl_cache/2` was expired. This is called `cleanup`.
- If the cache is full and also the newest cache entry is expired, all the entries will be deleted. This called `empty`.

### avg algorithm complexity:
- cache put: O(1)
- cache get: O(1)
- replacement: O(1)
- cleanup: O(N/2)
- empty: O(N)

### CODE REVIEW

File: etc/emqx.conf [hunk 1]
```diff
@@ -434,16 +434,22 @@ acl_nomatch = allow
 ## Value: File Name
 acl_file = {{ platform_etc_dir }}/acl.conf
 
-## Whether to enable ACL cache for publish.
+## The ACL cache size
+## The maximum count of ACL entries allowed for a client.
 ##
-## Value: on | off
-enable_acl_cache = on
+## Value 0 disables ACL cache
```
Comment (emqplus) [inline etc/emqx.conf]: Value should be greater than 0, for there is 'enable_acl_cache' option.

File: src/emqx.app.src [hunk 2]
```diff
@@ -3,7 +3,8 @@
               {vsn,"3.0"},
               {modules,[]},
               {registered,[emqx_sup]},
-              {applications,[kernel,stdlib,jsx,gproc,gen_rpc,lager,esockd,cowboy]},
+              {applications,[kernel,stdlib,jsx,gproc,gen_rpc,lager,esockd,cowboy,
+                             minirest]},
```
Comment (emqplus) [inline src/emqx.app.src]: Remove minirest

File: src/emqx_access_control.erl [hunk 3]
```diff
@@ -82,16 +81,19 @@ authenticate(Client, Password, [{Mod, State, _Seq} | Mods]) ->
...
-
-check_acl(#client{zone = Zone}, _PubSub, _Topic, []) ->
-    emqx_zone:get_env(Zone, acl_nomatch, deny);
-
-check_acl(Client, PubSub, Topic, [{Mod, State, _Seq}|AclMods]) ->
-    case Mod:check_acl({Client, PubSub, Topic}, State) of
-        allow  -> allow;
-        deny   -> deny;
-        ignore -> check_acl(Client, PubSub, Topic, AclMods)
+    CacheEnabled = (emqx_acl_cache:get_cache_max_size() =/= 0),
```
Comment (emqplus) [inline src/emqx_access_control.erl]: Use 'enable_acl_cache' option

File: src/emqx_access_control.erl [hunk 4]
```diff
@@ -193,6 +189,15 @@ terminate(_Reason, _State) ->
 code_change(_OldVsn, State, _Extra) ->
     {ok, State}.
 
+check_acl_from_plugins(#client{zone = Zone}, _PubSub, _Topic, []) ->
```
Comment (emqplus) [inline src/emqx_access_control.erl]: do_check_acl?

File: src/emqx_access_control.erl [hunk 5]
```diff
@@ -193,6 +189,15 @@ terminate(_Reason, _State) ->
...
+    emqx_zone:get_env(Zone, acl_nomatch, deny);
+check_acl_from_plugins(Client, PubSub, Topic, [{Mod, State, _Seq}|AclMods]) ->
```
Comment (emqplus) [inline src/emqx_access_control.erl]: do_check_acl?

**Comment by terry-xiaoyu** [comment]
updated some test cases

**Pull request merged**